Formula/r/rtabmap.rb

```diff
@@ -0,0 +1,55 @@
+class Rtabmap < Formula
+  desc "Visual and LiDAR SLAM library and standalone application"
+  homepage "https://introlab.github.io/rtabmap"
+  url "https://github.com/introlab/rtabmap/archive/refs/tags/0.21.3-noetic.tar.gz"
+  sha256 "7737018b3faa0b8887e79e057e3d8cf86cd9e2b3e8fe164e95dbf9585eabce8e"
+  license "BSD-3-Clause"
+  head "https://github.com/introlab/rtabmap.git", branch: "master"
+
+  depends_on "cmake" => :build
+  depends_on "g2o"
+  depends_on "librealsense"
+  depends_on "octomap"
+  depends_on "opencv"
+  depends_on "pcl"
+  depends_on "pdal"
+
+  def install
+    args = %W[
+      -DCMAKE_INSTALL_RPATH=#{rpath}
+    ]
+    system "cmake", "-S", ".", "-B", "build", *std_cmake_args, *args
+    system "cmake", "--build", "build"
+    system "cmake", "--install", "build"
+  end
+
+  test do
+    # Test if main app is built and can be launched
+    output = shell_output("#{bin}/rtabmap --version")
+    assert_match "RTAB-Map:               #{version}", output
+
+    # Test c++ library
+    ENV.delete "CPATH" # error xcode headers missing
```

Comment:
Does everyone need this?

Reply:
On my system yes (M1 - sonoma), not sure for everyone. There is something slightly different in build environment that I cannot see between the build and test phases. I added a new [commit](https://github.com/Homebrew/homebrew-core/pull/162576/commits/47b572f08879a05f938c664f9f62dcfbb284fe11) to add `depends_on` cmake for test, but it didn't help. The current workaround was inspired by https://github.com/Homebrew/homebrew-core/blob/7a51bbf301f5b4705d0171a8ec3a1edd52fb7233/Formula/p/pcl.rb#L116

For me, there are many xcode related errors (missing headers), but here is the first one:
```
/Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk/usr/include/c++/v1/cmath:320:5: error: <cmath> tried
 including <math.h> but didn't find libc++'s <math.h> header.           This usually means that your header search 
paths are not configured properly.           The header search paths should contain the C++ Standard Library 
headers before           any C Standard Library, and you are probably using compiler flags that make that           not
 be the case.
```

The compilation command not working:
```
/Library/Developer/CommandLineTools/usr/bin/clang++ -DBOOST_ATOMIC_DYN_LINK -DBOOST_ATOMIC_NO_LIB
 -DBOOST_FILESYSTEM_DYN_LINK -DBOOST_FILESYSTEM_NO_LIB -DBOOST_IOSTREAMS_DYN_LINK -
DBOOST_IOSTREAMS_NO_LIB -DBOOST_SERIALIZATION_DYN_LINK -DBOOST_SERIALIZATION_NO_LIB -
DBOOST_SYSTEM_DYN_LINK -DBOOST_SYSTEM_NO_LIB -DEIGEN_HAS_CXX17_OVERALIGN=0 -DQT_CORE_LIB
 -DQT_GUI_LIB -DQT_NO_DEBUG -DQT_OPENGLWIDGETS_LIB -DQT_OPENGL_LIB -DQT_WIDGETS_LIB -
Dkiss_fft_scalar=double -isystem /opt/homebrew/include/eigen3 -isystem /opt/homebrew/Cellar/rtabmap/0.21.3/
include/rtabmap-0.21 -isystem /opt/homebrew/Cellar/opencv/4.9.0_2/include/opencv4 -isystem /opt/homebrew/
include/pcl-1.14 -isystem /opt/homebrew/include -isystem /opt/homebrew/opt/libpcap/include -isystem /Library/
Developer/CommandLineTools/SDKs/MacOSX14.sdk/usr/include -isystem /opt/homebrew/include/vtk-9.2 -isystem /
opt/homebrew/include/vtk-9.2/vtknlohmannjson/include -isystem /opt/homebrew/include/vtk-9.2/vtkfreetype/
include -isystem /opt/homebrew/lib/QtOpenGL.framework/Headers -iframework /opt/homebrew/lib -isystem /opt/
homebrew/lib/QtCore.framework/Headers -isystem /opt/homebrew/share/qt/mkspecs/macx-clang -isystem /opt/
homebrew/lib/QtGui.framework/Headers -isystem /opt/homebrew/lib/QtWidgets.framework/Headers -isystem /opt/
homebrew/lib/QtOpenGLWidgets.framework/Headers -Os -w -pipe  -mmacosx-version-min=14 -isysroot/Library/
Developer/CommandLineTools/SDKs/MacOSX14.sdk -std=gnu++17 -arch arm64 -isysroot /Library/Developer/
CommandLineTools/SDKs/MacOSX14.sdk -ffloat-store -MD -MT CMakeFiles/test.dir/test.cpp.o -MF CMakeFiles/
test.dir/test.cpp.o.d -o CMakeFiles/test.dir/test.cpp.o -c /tmp/rtabmap-test-20240214-2050-b8d70e/test.cpp
```

The compilation command working (with `ENV.delete`):
```bash
/Library/Developer/CommandLineTools/usr/bin/clang++ -DBOOST_ATOMIC_DYN_LINK -DBOOST_ATOMIC_NO_LIB 
-DBOOST_FILESYSTEM_DYN_LINK -DBOOST_FILESYSTEM_NO_LIB -DBOOST_IOSTREAMS_DYN_LINK -
DBOOST_IOSTREAMS_NO_LIB -DBOOST_SERIALIZATION_DYN_LINK -DBOOST_SERIALIZATION_NO_LIB -
DBOOST_SYSTEM_DYN_LINK -DBOOST_SYSTEM_NO_LIB -DEIGEN_HAS_CXX17_OVERALIGN=0 -DQT_CORE_LIB
 -DQT_GUI_LIB -DQT_NO_DEBUG -DQT_OPENGLWIDGETS_LIB -DQT_OPENGL_LIB -DQT_WIDGETS_LIB -
Dkiss_fft_scalar=double -isystem /opt/homebrew/include/eigen3 -isystem /opt/homebrew/Cellar/rtabmap/0.21.3/
include/rtabmap-0.21 -isystem /opt/homebrew/Cellar/opencv/4.9.0_2/include/opencv4 -isystem /opt/homebrew/
include/pcl-1.14 -isystem /opt/homebrew/include -isystem /opt/homebrew/opt/libpcap/include -isystem /opt/
homebrew/include/vtk-9.2 -isystem /opt/homebrew/include/vtk-9.2/vtknlohmannjson/include -isystem /opt/
homebrew/include/vtk-9.2/vtkfreetype/include -isystem /opt/homebrew/lib/QtOpenGL.framework/Headers -
iframework /opt/homebrew/lib -isystem /opt/homebrew/lib/QtCore.framework/Headers -isystem /opt/homebrew/
share/qt/mkspecs/macx-clang -isystem /opt/homebrew/lib/QtGui.framework/Headers -isystem /opt/homebrew/lib/
QtWidgets.framework/Headers -isystem /opt/homebrew/lib/QtOpenGLWidgets.framework/Headers -Os -w -pipe  -
mmacosx-version-min=14 -isysroot/Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk -std=gnu++17 -
arch arm64 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk -ffloat-store -MD -MT 
CMakeFiles/test.dir/test.cpp.o -MF CMakeFiles/test.dir/test.cpp.o.d -o CMakeFiles/test.dir/test.cpp.o -c /tmp/
rtabmap-test-20240214-1399-n4wkj8/test.cpp
```

Note also that if I go directly in the `/tmp/rtabmap-test-20240214-1399-n4wkj8`directory, remove the `CMakeCache.txt` and do simply `cmake .`, `make`, it works.

I'll give a try today on a Mac Intel at work and see if it is just my computer the issue.

Reply:
Confirmed it was also happening on CI: https://github.com/Homebrew/homebrew-core/actions/runs/7983978459/job/21800145992?pr=162576

I added `ENV.delete "CPATH"` back to fix this issue.

Reply:
I searched for all other Formulas using `ENV.delete "CPATH"`:
https://github.com/search?q=repo%3AHomebrew%2Fhomebrew-core%20ENV.delete%20%22CPATH%22&type=code

Here some related discussions:
 * https://github.com/ziglang/zig/issues/10377#issuecomment-1000895055
   
> Based on CI at https://github.com/Homebrew/homebrew-core/pull/91763, it seems that it's the CPATH setting that's the culprit. My guess is that it ends up mixing headers from Xcode and the CLT and that breaks things.

 * In qt formula they added it in this commit https://github.com/matlabbe/homebrew-core/commit/c844c6d47a9a9189500b7962a94be18fafcafdef, seems related to MacOS versions since Catalina:
  ```
  ENV.delete "CPATH" if OS.mac? && MacOS.version > :mojave
  ```
  
For all other formulas, I see it often when they use Qt with qmake or cmake.

Reply:
Found some PRs mentionning it:
https://github.com/search?q=repo%3AHomebrew%2Fhomebrew-core+ENV.delete+%22CPATH%22&type=pullrequests

* https://github.com/Homebrew/homebrew-core/pull/75166#issuecomment-899602384
>My local builds seem to work as I have mentioned https://github.com/root-project/root/issues/7881#issuecomment-850791642. The failing logs of this PR still have -I/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include which I think makes it fail.

> Have you tried adding a depends_on :xcode on Catalina and earlier? depends_on xcode: :build might also work. This will make brew stop trying to use the CLT. This is otherwise a lot more involved than just the content of std_cmake_args.

* https://github.com/Homebrew/homebrew-core/pull/76670#issuecomment-832922677
> Yes, that would make it switch from the Xcode 10.15 SDK to the CLT 10.14 SDK. It's also why you suddenly needed a ENV.delete "CPATH" in your test (qmake doesn't like CLT SDKs very much).

 * https://github.com/Homebrew/homebrew-core/pull/66422#issuecomment-740166386

>  Work around "error: no member named 'signbit' in the global namespace" 
 ENV.delete("SDKROOT") if DevelopmentTools.clang_build_version >= 900 

* https://github.com/Homebrew/homebrew-core/pull/39181#issue-436055079
> Since the 2nd failure seems to be related to C header files, I added ENV.delete("CPATH"), and that seems to be enough.

* https://github.com/Homebrew/homebrew-core/pull/794#issuecomment-220075564
> @apjanke Do we have any way to require CLT be installed? Even with ENV.delete("SDKROOT") or with https://github.com/Homebrew/brew/pull/251 baresip will fail if only Xcode is installed.

Reply:
I'm not concerned about this in a test but would like to see a better comment and more finely scoped if possible.

```suggestion
    # Required to avoid missing Xcode headers 
    # https://github.com/Homebrew/homebrew-core/pull/162576/files#r1489824628
    ENV.delete "CPATH" if OS.mac? && !MacOS::CLT.installed?
```

or similar